backend/service/sourcegraph/sourcegraph.go

```diff
@@ -0,0 +1,158 @@
+package sourcegraph
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+
+	sourcegraphv1cfg "github.com/lyft/clutch/backend/api/config/service/sourcegraph/v1"
+	sourcegraphv1 "github.com/lyft/clutch/backend/api/sourcegraph/v1"
+	"github.com/lyft/clutch/backend/service"
+)
+
+const Name = "clutch.service.sourcegraph"
+
+type Service interface {
+	CompareCommits(context.Context, *sourcegraphv1.CompareCommitsRequest) (*sourcegraphv1.CompareCommitsResponse, error)
+}
+
+type QraphQLQuery struct {
+	Query string `json:"query"`
+}
+
+type client struct {
+	config *sourcegraphv1cfg.Config
+	log    *zap.Logger
+	scope  tally.Scope
+
+	sgURL  *url.URL
+	client *http.Client
+}
+
+func New(cfg *any.Any, log *zap.Logger, scope tally.Scope) (service.Service, error) {
+	sgConfig := &sourcegraphv1cfg.Config{}
+	err := cfg.UnmarshalTo(sgConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	sgURL, err := url.Parse(fmt.Sprintf("%s%s", sgConfig.Host, "/.api/graphql"))
+	if err != nil {
+		log.Error("unable to parse sourcegraph host", zap.Error(err))
+		return nil, err
+	}
+
+	// setup http client with auth token
+	ctx := context.Background()
+	sgClient := oauth2.NewClient(ctx, oauth2.StaticTokenSource(&oauth2.Token{
+		AccessToken: sgConfig.Token,
+		TokenType:   "token",
+	}))
+
+	return &client{
+		config: sgConfig,
+		log:    log,
+		scope:  scope,
+
+		sgURL:  sgURL,
+		client: sgClient,
+	}, nil
+}
+
+func (c *client) CompareCommits(ctx context.Context, req *sourcegraphv1.CompareCommitsRequest) (*sourcegraphv1.CompareCommitsResponse, error) {
+	compareCommitQuery := fmt.Sprintf(`
```

Comment:
did you look at https://github.com/shurcooL/graphql?

it provides a much more structured way of handling graphql in go rather than having a query and then redefining the response types in separate structs. but it might be more complicated to get started with. curious what your thoughts are though.

Reply:
Alright check it out, looking much cleaner.